core/incremental/expr_compiler.rs

```diff
@@ -0,0 +1,441 @@
+        match self {
+            TrivialExpression::Column(idx) => values.get(*idx).cloned().unwrap_or(Value::Null),
+            TrivialExpression::Immediate(val) => val.clone(),
+            TrivialExpression::Binary { left, op, right } => {
+                let left_val = left.evaluate(values);
+                let right_val = right.evaluate(values);
+
+                // Only perform operations on same-type operands
+                match op {
+                    Operator::Add => match (&left_val, &right_val) {
+                        (Value::Integer(a), Value::Integer(b)) => Value::Integer(a + b),
+                        (Value::Float(a), Value::Float(b)) => Value::Float(a + b),
+                        (Value::Null, _) | (_, Value::Null) => Value::Null,
+                        _ => panic!("Type mismatch in trivial expression: {left_val:?} + {right_val:?}. This is a bug in trivial expression validation."),
+                    },
+                    Operator::Subtract => match (&left_val, &right_val) {
+                        (Value::Integer(a), Value::Integer(b)) => Value::Integer(a - b),
+                        (Value::Float(a), Value::Float(b)) => Value::Float(a - b),
+                        (Value::Null, _) | (_, Value::Null) => Value::Null,
+                        _ => panic!("Type mismatch in trivial expression: {left_val:?} - {right_val:?}. This is a bug in trivial expression validation."),
+                    },
+                    Operator::Multiply => match (&left_val, &right_val) {
+                        (Value::Integer(a), Value::Integer(b)) => Value::Integer(a * b),
+                        (Value::Float(a), Value::Float(b)) => Value::Float(a * b),
+                        (Value::Null, _) | (_, Value::Null) => Value::Null,
+                        _ => panic!("Type mismatch in trivial expression: {left_val:?} * {right_val:?}. This is a bug in trivial expression validation."),
+                    },
+                    Operator::Divide => match (&left_val, &right_val) {
+                        (Value::Integer(a), Value::Integer(b)) => {
+                            if *b != 0 {
+                                Value::Integer(a / b)
+                            } else {
+                                Value::Null
+                            }
+                        }
+                        (Value::Float(a), Value::Float(b)) => {
+                            if *b != 0.0 {
+                                Value::Float(a / b)
+                            } else {
+                                Value::Null
+                            }
+                        }
+                        (Value::Null, _) | (_, Value::Null) => Value::Null,
+                        _ => panic!("Type mismatch in trivial expression: {left_val:?} / {right_val:?}. This is a bug in trivial expression validation."),
+                    },
+                    _ => panic!("Unsupported operator in trivial expression: {op:?}"),
```

Comment:
Good not to swallow the fallthrough case, but maybe make this function return a Result?

Reply:
humm... The thing is that if this is not a trivial operation, then we should not have tried to construct a trivial operation out of it. This is just a violation about our assumptions about the code, which is usually what we use asserts for.